### PR TITLE
Src: Add global usings

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Cleanup will be a separate PR once this one is merged

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CFG/SonarAnalyzer.CFG.csproj [hunk 1]
```diff
@@ -22,19 +24,11 @@
...
-
-  <ItemGroup>
     <Compile Update="ShimLayer\*.cs">
       <SonarQubeExclude>true</SonarQubeExclude>
     </Compile>
   </ItemGroup>
 
-  <PropertyGroup>
-    <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
-    <CompilerGeneratedFilesOutputPath>$(MSBuildProjectDirectory)\Lightup\.generated</CompilerGeneratedFilesOutputPath>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/SonarAnalyzer.CFG.csproj:35]: Just moved up, no need to persist two sets of `PropertyGroup`

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Sonar/SymbolicValues/NullableSymbolicValue.cs [hunk 2]
```diff
@@ -51,7 +51,7 @@ public override IEnumerable<ProgramState> TrySetConstraint(SymbolicConstraint co
                 return TrySetConstraint(optionalConstraint, programState);
             }
 
-            var oldConstraint = programState.Constraints.GetValueOrDefault(this)?.GetConstraintOrDefault<NullableConstraint>();
+            var oldConstraint = ((IDictionary<SymbolicValue, SymbolicValueConstraints>)programState.Constraints).GetValueOrDefault(this)?.GetConstraintOrDefault<NullableConstraint>();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Sonar/SymbolicValues/NullableSymbolicValue.cs:54]: I don't care about this one, it's the old SE engine. See other comments below.

File: analyzers/src/SonarAnalyzer.Common/Common/SecondaryLocation.cs [hunk 3]
```diff
@@ -59,6 +59,6 @@ public static class SecondaryLocationHelper
             diagnostic.AdditionalLocations.Count <= index
                     ? throw new ArgumentOutOfRangeException(nameof(index))
                     : new SecondaryLocation(diagnostic.AdditionalLocations[index],
-                                            diagnostic.Properties.GetValueOrDefault(index.ToString()));
+                                            ((IDictionary<string, string>)diagnostic.Properties).GetValueOrDefault(index.ToString()));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Common/SecondaryLocation.cs:62]: There's unfortunate clash in extension names between IDictonary and IImmutableDictionary -> penalty for overusing extensions by library makers 😞

File: analyzers/src/Directory.Build.targets [hunk 4]
```diff
@@ -7,6 +7,18 @@
     <BinariesFolder>$(MSBuildThisFileDirectory)\..\packaging\binaries\</BinariesFolder>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Using Include="System" />
+    <Using Include="System.Collections.Generic" />
+    <Using Include="System.Collections.Immutable" />
+    <Using Include="System.Diagnostics" />
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/Directory.Build.targets:14]: What was the way you decided which namespaces to include? For example `System.Diagnostics` has only 12 usages through our codebase, and I am not sure it is worth it adding here.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/Directory.Build.targets:14]: For developer comfort reasons (speed of development). It's one of the namespaces that are in VB by default. It's a veeery practical to be able to type just `Debugger.Break()` or `Debugger.Launch()` anywhere, anytime.

**Comment by Corniel** [comment]
I'm just curious: why did you go for global usings in the project file? And not in a code file? (For instance "Properties/GlobalUsings.cs")

**Comment by pavel-mikula-sonarsource** [comment]
Most of them can be shared between all projects, see `Directory.Build.targets` file.

Also putting this global configuration into a random file anywhere in the project was a very bad design idea from MS. They should have done just the same thing that was working for VB.NET already for many years.

Up to this day, they even don't have an API to read that pain 😒

**Comment by pavel-mikula-sonarsource** [comment]
The problem I see is that nothing prevents devs from putting one more keyword in another random file that would change the whole project.
I haven't tried the aliases either, we don't use them that often.

**Comment by Corniel** [comment]
> The problem I see is that nothing prevents devs from putting one more keyword in another random file that would change the whole project.

I know some guys who build awesome static code analysis rules that could write a rule that prevents just that. :)

**Comment by pavel-mikula-sonarsource** [comment]
> I know some guys who build awesome static code analysis rules that could write a rule that prevents just that. :)

Let's see
S0001 don't use this feature and move it to your project file 
S0002 don't use global alias, just give it a good name instead :trollface: 
😅

**Comment by Corniel** [comment]
> > I know some guys who build awesome static code analysis rules that could write a rule that prevents just that. :)
> 
> Let's see [S0001](https://sonarsource.github.io/rspec/#/rspec/S0001/csharp) don't use this feature and move it to your project file [S0002](https://sonarsource.github.io/rspec/#/rspec/S0002/csharp) don't use global alias, just give it a good name instead :trollface: 😅

We use a generic struct `Id<T>`. there aliases help. But happy trolling today. :)

**Comment by pavel-mikula-sonarsource** [comment]
Now a bit more seriously, if you have ideas for practical rules (not neccessary SonarWay) please open an issues.
I can think of rules like "keep them all in one file", "keep them only in well-known file" (where well-known would be a parameter)

**Comment by Corniel** [comment]
> Now a bit more seriously, if you have ideas for practical rules (not neccessary SonarWay) please open an issues. I can think of rules like "keep them all in one file", "keep them only in well-known file" (where well-known would be a parameter)

I have. Unfortunately, a bit like: [Classes should be sealed by default](https://community.sonarsource.com/t/new-rule-classes-should-be-sealed-final-by-default/70271/4), it is more based on personal preference.

File: analyzers/src/Directory.Build.targets [hunk 5]
```diff
@@ -7,6 +7,18 @@
     <BinariesFolder>$(MSBuildThisFileDirectory)\..\packaging\binaries\</BinariesFolder>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Using Include="System" />
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/Directory.Build.targets:11]: WDYT about moving the shared usings from tests and src to analyzer/Directory.Build.targets?

**Review by csaba-sagi-sonarsource — COMMENTED** [review]
Also now with these updates the global usings are in build.target files for the "main" code while it is in the .csproj for the test code. Woudl you mind moving the test specific global usings to their Directory.Build.targets file? Just for consistency.

**Review by csaba-sagi-sonarsource — COMMENTED** [review]
My overall felling is that now we have global usings all over the place. Some are in the .csproj files while others are in the Directory.Build.target files. So, if you are looking for one it is difficult to find where it comes from or where you should add one if you want. Maybe putting them into some separate well-named files would not do any harm at the end of the day.

**Comment by pavel-mikula-sonarsource** [comment]
They'll be all over the place in any case, because one file cannot add all. It fails the build if nonexistent using is added to a project.

**Comment by pavel-mikula-sonarsource** [comment]
Test-specific usings cannot go to `Directory.Build.targets`, because it would fail the CBDE build. As those projects do not contain test framework references.

**Comment by pavel-mikula-sonarsource** [comment]
After the review update, all global stuff is in a single file. And each project-specific using are in respective project file. That's reasonable setup.

**Comment by csaba-sagi-sonarsource** [comment]
> Test-specific usings cannot go to `Directory.Build.targets`, because it would fail the CBDE build. As those projects do not contain test framework references.

I did not mean to add test-specific usings, sorry if it was not clear what I wanted to suggest. I wanted to suggest exactly what you did, usings like system and collections to add to a separate file and use them for tests too.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**